src/mesh/p4est_mesh.jl

```diff
@@ -199,15 +195,23 @@ function P4estMesh(meshfile::String;
                                                 n_trees)
   calc_tree_node_coordinates!(tree_node_coordinates, nodes, mapping, vertices, tree_to_vertex)
 
-  p4est = p4est_new_ext(0, conn, 0, initial_refinement_level, true, 0, C_NULL, C_NULL)
-  ghost = p4est_ghost_new(p4est, P4EST_CONNECT_FACE)
-  p4est_mesh = p4est_mesh_new(p4est, ghost, P4EST_CONNECT_FACE)
+  # Use Int-Vector of size 2 as quadrant user data
+  p4est = p4est_new_ext(0, conn, 0, initial_refinement_level, true, 2 * sizeof(Int), C_NULL, C_NULL)
 
   # There's no simple and generic way to distinguish boundaries. Name all of them :all.
   boundary_names = fill(:all, 4, n_trees)
 
-  return P4estMesh{NDIMS, RealT, NDIMS+2}(p4est, p4est_mesh, tree_node_coordinates, nodes,
+  mesh = P4estMesh{NDIMS, RealT, NDIMS+2}(p4est, tree_node_coordinates, nodes,
                                           boundary_names, "", unsaved_changes)
+
+  # Destroy p4est structs when the mesh is garbage collected
+  finalizer(mesh) do
+    conn = mesh.p4est.connectivity
+    p4est_destroy(mesh.p4est)
+    p4est_connectivity_destroy(conn)
+  end
```

Comment:
Why does this cause these awkward, super long error messages?

Reply:
Do you get similar errors locally?

Reply:
If this turns out to be difficult to debug locally, I propose to revert this change back to the version you had before and put it on the TODO list of things to fix before merging `dev` into `main`.

Reply:
I get the same error locally. I have absolutely no idea how to even start debugging, I just get this non-descriptive error 
```
error in running finalizer: MethodError(f=Trixi.var"#735", args=(Trixi.P4estMesh{2, Float64, 4}(p4est=0x0000000098804cd0, tree_node_coordinates=[...], current_filename="out\mesh.h5", unsaved_changes=false),), world=0x0000000000007526)
jl_method_error_bare at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:1812
jl_method_error at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:1830
jl_lookup_generic_ at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:2400
jl_apply_generic at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:2415
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1703 [inlined]
run_finalizer at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:278
jl_gc_run_finalizers_in_list at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:365
run_finalizers at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:394 [inlined]
run_finalizers at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:372
jl_gc_run_pending_finalizers at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:404
jl_mutex_unlock at /cygdrive/c/buildbot/worker/package_win64/build/src\locks.h:136 [inlined]
jl_generate_fptr at /cygdrive/c/buildbot/worker/package_win64/build/src\jitlayers.cpp:361
jl_compile_method_internal at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:1970
jl_compile_method_internal at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:1924 [inlined]
_jl_invoke at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:2229 [inlined]
jl_apply_generic at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:2419
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1703 [inlined]
do_call at /cygdrive/c/buildbot/worker/package_win64/build/src\interpreter.c:115
eval_value at /cygdrive/c/buildbot/worker/package_win64/build/src\interpreter.c:204
eval_stmt_value at /cygdrive/c/buildbot/worker/package_win64/build/src\interpreter.c:155 [inlined]
eval_body at /cygdrive/c/buildbot/worker/package_win64/build/src\interpreter.c:576
jl_interpret_toplevel_thunk at /cygdrive/c/buildbot/worker/package_win64/build/src\interpreter.c:670
top-level scope at c:\Users\Erik\git\Trixi.jl\examples\2d\elixir_advection_p4est_non_conforming_flag_unstructured.jl:31
jl_toplevel_eval_flex at /cygdrive/c/buildbot/worker/package_win64/build/src\toplevel.c:877
jl_toplevel_eval_flex at /cygdrive/c/buildbot/worker/package_win64/build/src\toplevel.c:825
jl_toplevel_eval at /cygdrive/c/buildbot/worker/package_win64/build/src\toplevel.c:886 [inlined]
jl_toplevel_eval_in at /cygdrive/c/buildbot/worker/package_win64/build/src\toplevel.c:929
eval at .\boot.jl:360 [inlined]
include_string at .\loading.jl:1094
include_string at .\loading.jl:1104
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1703 [inlined]
jl_f__call_latest at /cygdrive/c/buildbot/worker/package_win64/build/src\builtins.c:714
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1703 [inlined]
do_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\builtins.c:670
#invokelatest#2 at .\essentials.jl:708
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1703 [inlined]
do_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\builtins.c:670
invokelatest at .\essentials.jl:706
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1703 [inlined]
do_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\builtins.c:670
#inlineeval#63 at c:\Users\Erik\.vscode\extensions\julialang.language-julia-1.2.2\scripts\packages\VSCodeServer\src\eval.jl:204
inlineeval##kw at c:\Users\Erik\.vscode\extensions\julialang.language-julia-1.2.2\scripts\packages\VSCodeServer\src\eval.jl:202 [inlined]
#58 at c:\Users\Erik\.vscode\extensions\julialang.language-julia-1.2.2\scripts\packages\VSCodeServer\src\eval.jl:156
withpath at c:\Users\Erik\.vscode\extensions\julialang.language-julia-1.2.2\scripts\packages\VSCodeServer\src\repl.jl:177
#57 at c:\Users\Erik\.vscode\extensions\julialang.language-julia-1.2.2\scripts\packages\VSCodeServer\src\eval.jl:154
hideprompt at c:\Users\Erik\.vscode\extensions\julialang.language-julia-1.2.2\scripts\packages\VSCodeServer\src\repl.jl:36
#56 at c:\Users\Erik\.vscode\extensions\julialang.language-julia-1.2.2\scripts\packages\VSCodeServer\src\eval.jl:122
unknown function (ip: 000000006c57cec3)
with_logstate at .\logging.jl:491
with_logger at .\logging.jl:603 [inlined]
#55 at c:\Users\Erik\.vscode\extensions\julialang.language-julia-1.2.2\scripts\packages\VSCodeServer\src\eval.jl:194
unknown function (ip: 000000006c57ca93)
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1703 [inlined]
jl_f__call_latest at /cygdrive/c/buildbot/worker/package_win64/build/src\builtins.c:714
#invokelatest#2 at .\essentials.jl:708 [inlined]
invokelatest at .\essentials.jl:706
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1703 [inlined]
do_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\builtins.c:670
macro expansion at c:\Users\Erik\.vscode\extensions\julialang.language-julia-1.2.2\scripts\packages\VSCodeServer\src\eval.jl:34 [inlined]
#53 at .\task.jl:411
unknown function (ip: 0000000061080593)
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1703 [inlined]
start_task at /cygdrive/c/buildbot/worker/package_win64/build/src\task.c:839

Reply:
Running a function
```julia
function destroy(mesh)
    conn = mesh.p4est.connectivity
    p4est_destroy(mesh.p4est)
    p4est_connectivity_destroy(conn)
end
```
manually after a simulation works perfectly fine. I tried it hundreds of times. The finalizer works often but fails sometimes.

Reply:
I.e., if you define it as a normal function instead of a closure?

Reply:
If that works, I would prefer moving the this to an inner constructor such that you don't need to repeat the same code twice (or even more often).

Reply:
That actually works, thanks!
The only line that's called twice is
```julia
finalizer(destroy_mesh, mesh)
```

Reply:
Yeah, I would nevertheless like to have it in an inner constructor. We will probably get more outer (convenience) constructors and it's pretty easy to forget this line, introducing memory bugs etc.